DataFormats/Portable/interface/PortableDeviceProduct.h

```diff
@@ -0,0 +1,72 @@
+#ifndef DataFormats_Portable_interface_PortableDeviceProduct_h
+#define DataFormats_Portable_interface_PortableDeviceProduct_h
+
+#include <cassert>
+#include <optional>
+#include <type_traits>
+
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+
+// generic SoA-based product in device memory
+template <typename T, typename TDev, typename = std::enable_if_t<alpaka::isDevice<TDev>>>
+class PortableDeviceProduct {
+  static_assert(not std::is_same_v<TDev, alpaka_common::DevHost>,
+                "Use PortableHostProduct<T> instead of PortableDeviceProduct<T, DevHost>");
+
+public:
+  using Product = T;
+  using Buffer = cms::alpakatools::device_buffer<TDev, Product>;
+  using ConstBuffer = cms::alpakatools::const_device_buffer<TDev, Product>;
+
+  PortableDeviceProduct() = default;
+
+  PortableDeviceProduct(TDev const& device)
+      // allocate global device memory
+      : buffer_{cms::alpakatools::make_device_buffer<Product>(device)} {
+    assert(reinterpret_cast<uintptr_t>(buffer_->data()) % alignof(Product) == 0);
+  }
+
+  template <typename TQueue, typename = std::enable_if_t<alpaka::isQueue<TQueue>>>
+  PortableDeviceProduct(TQueue const& queue)
+      // allocate global device memory with queue-ordered semantic
+      : buffer_{cms::alpakatools::make_device_buffer<Product>(queue)} {
+    assert(reinterpret_cast<uintptr_t>(buffer_->data()) % alignof(Product) == 0);
+  }
+
+  // non-copyable
+  PortableDeviceProduct(PortableDeviceProduct const&) = delete;
+  PortableDeviceProduct& operator=(PortableDeviceProduct const&) = delete;
+
+  // movable
+  PortableDeviceProduct(PortableDeviceProduct&&) = default;
+  PortableDeviceProduct& operator=(PortableDeviceProduct&&) = default;
+
+  // default destructor
+  ~PortableDeviceProduct() = default;
+
+  // access the product
+  Product& value() { return *buffer_->data(); }
+  Product const& value() const { return *buffer_->data(); }
+
+  Product* data() { return buffer_->data(); }
+  Product const* data() const { return buffer_->data(); }
+
+  Product& operator*() { return *buffer_->data(); }
+  Product const& operator*() const { return *buffer_->data(); }
+
+  Product* operator->() { return buffer_->data(); }
+  Product const* operator->() const { return buffer_->data(); }
+
+  // access the buffer
+  Buffer buffer() { return *buffer_; }
```

Comment:
Could this be
```suggestion
  Buffer& buffer() { return *buffer_; }
```
(although I see the same pattern in `PortableDeviceCollection`)

Reply:
No, it really should be a copy.
Otherwise one could modify the internals of the buffer.

Reply:
Ok